### DISCUSSION TITLE
Token refresh is not working for me

### DISCUSSION THREAD

**Discussion by HansKappert** [tdorssers/TeslaPy#130] [Q&A]
I have obtained a token (cache.json) via command
`python3 cli.py -e <my mail address>`
I copied the cache.json to the orange zero headless computer where it functions for a while (can interact with my Tesla).
After a while (hours) I get this exception:
`Exception during setting of the current: Traceback (most recent call last): File "/home/pi/venv/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 483, in request url, headers, data = self._client.add_token( File "/home/pi/venv/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 217, in add_token raise TokenExpiredError() oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) During handling of the above exception, another exception occurred: Traceback (most recent call last): File "/home/pi/net_balancer-main/src/service/tesla_energy_consumer.py", line 129, in start_consuming self.__set_charge_current(new_charging_current) File "/home/pi/net_balancer-main/src/service/tesla_energy_consumer.py", line 278, in __set_charge_current self.vehicle.command("CHARGING_AMPS",charging_amps=amps) File "/home/pi/net_balancer-main/src/teslapy/__init__.py", line 711, in command response = self.api(name, **kwargs).get('response') File "/home/pi/net_balancer-main/src/teslapy/__init__.py", line 488, in api return self.tesla.api(name, {'vehicle_id': self['id_s']}, **kwargs) File "/home/pi/net_balancer-main/src/teslapy/__init__.py", line 369, in api return self.request(endpoint['TYPE'], uri, serialize, File "/home/pi/net_balancer-main/src/teslapy/__init__.py", line 150, in request response = super(Tesla, self).request(method, url, **kwargs) File "/home/pi/venv/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 502, in request token = self.refresh_token( File "/home/pi/net_balancer-main/src/teslapy/__init__.py", line 249, in refresh_token super(Tesla, self).refresh_token(token_url, **kwargs) File "/home/pi/venv/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 452, in refresh_token self.token = self._client.parse_request_body_response(r.text, scope=self.scope) File "/home/pi/venv/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 427, in parse_request_body_response self.token = parse_token_response(body, scope=scope) File "/home/pi/venv/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 441, in parse_token_response validate_token_parameters(params) File "/home/pi/venv/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 451, in validate_token_parameters raise MissingTokenError(description="Missing access token parameter.") oauthlib.oauth2.rfc6749.errors.MissingTokenError: (missing_token) Missing access token parameter. 
`
It looks as if the auto refresh is not happening. 
I have no clue what I am doing wrong here. Does anybody have a suggestion?

**Comment by HansKappert** [comment]
Additional info. I turned on debug logging on the orangePI. It tries to refresh the token, but this is what happens:

`jun 05 21:46:53 orangepizero python3[17945]: 2023-06-05 21:46:53,676 - urllib3.connectionpool - DEBUG - https://auth.tesla.com:443 "POST /oauth2/v3/token HTTP/1.1" 403 286`
`jun 05 21:46:53 orangepizero python3[17945]: 2023-06-05 21:46:53,688 - requests_oauthlib.oauth2_session - DEBUG - Request to refresh token completed with status 403.`
`jun 05 21:46:53 orangepizero python3[17945]: 2023-06-05 21:46:53,693 - requests_oauthlib.oauth2_session - DEBUG - Response headers were {'Server': 'AkamaiGHost', 'Mime-Version': '1.0', 'Content-Type': 'text/html', 'Content-Length': '286', 'Expires': 'Mon, 05 Jun 2023 19:46:53 GMT', 'X-Reference-Error': '18.ead31702.1685994413.3bb9418', 'Date': 'Mon, 05 Jun 2023 19:46:53 GMT', 'Connection': 'close', 'Akamai-Request-BC': '[a=2.23.211.234,b=62624792,c=g,n=NL__SCHIPHOL,o=20940]', 'Permissions-Policy': 'interest-cohort=()'} and content <HTML><HEAD>`
`jun 05 21:46:53 orangepizero python3[17945]: <TITLE>Access Denied</TITLE>`
`jun 05 21:46:53 orangepizero python3[17945]: </HEAD><BODY>`
`jun 05 21:46:53 orangepizero python3[17945]: <H1>Access Denied</H1>`
`jun 05 21:46:53 orangepizero python3[17945]:`
`jun 05 21:46:53 orangepizero python3[17945]: You don't have permission to access "http&#58;&#47;&#47;auth&#46;tesla&#46;com&#47;oauth2&#47;v3&#47;token" on this server.<P>`
`jun 05 21:46:53 orangepizero python3[17945]: Reference&#32;&#35;18&#46;ead31702&#46;1685994413&#46;3bb9418`
`jun 05 21:46:53 orangepizero python3[17945]: </BODY>`
`jun 05 21:46:53 orangepizero python3[17945]: </HTML>`
`jun 05 21:46:53 orangepizero python3[17945]: .`
`jun 05 21:46:53 orangepizero python3[17945]: 2023-06-05 21:46:53,694 - requests_oauthlib.oauth2_session - DEBUG - Invoking 0 token response hooks.`
`jun 05 21:46:53 orangepizero python3[17945]: Error during initializing Tesla energy consumer: (missing_token) Missing access token parameter.`

**Comment by buehlerfabian** [comment]
I had the same problem and could solve it by going back to older versions of some of the python packages used by teslapy. Have a look at this issue:

https://github.com/tdorssers/TeslaPy/issues/128

Fabian

**Comment by HansKappert** [comment]
Hi Fabian,That sounds really promising. I will test this asap.Cheers,HansOp 6 jun. 2023 om 16:20 heeft fbphysik ***@***.***> het volgende geschreven:﻿
I had the same problem and could solve it by going back to older versions of some of the python packages used by teslapy. Have a look at this issue:
#128
Fabian

**Comment by HansKappert** [comment]
Works like a charm. Thanks a million!!!!!!

**Reply by SylvainGa** [reply]
Had the same problem and reverting the packages worked for me too. Are we locked at those version of the packages or will an updated version of teslapi support the newest packages?

**Reply by tdorssers** [reply]
Do we know which of those packages is the culprit?

**Comment by HansKappert** [comment]
I did not investigate that. Op 23 jun. 2023 om 12:59 heeft Tim Dorssers ***@***.***> het volgende geschreven:﻿
Do we know which of those packages is the culprit?

**Comment by buehlerfabian** [comment]
> Do we know which of those packages is the culprit?

I just checked:
Downgrading 'requests' to version 2.28.1 solves the problem. With this downgrade 'urllib3' automatically gets downgraded to version 1.26.16, so I can't say if the problem is within 'requests' or 'urllib3'.

The other downgrades ('certifi', 'charset-normalizer', 'websocket-client') I mentioned in #128  don't seem to be necessary.

**Comment by tdorssers** [comment]
I can't reproduce this. I tried using an expired token, to trigger auto refresh but I don't get a 403. Is it really needed to have a script running for over 8 hours to trigger this? I am looking for a minimal script to reproduce this

**Comment by buehlerfabian** [comment]
No, there is no long running script necessary to reproduce the error. My steps are:

1) Create a new virtual environment (using poetry).
2) Install the packages 'teslapy' and the packages that it depends on, using current pypi versions.
3) Use a 'cache.json' file with an expired token to trigger refresh.
3a) OR: Use no 'cache.json' file and go through the authentication process -> also triggers the error
4) Run the following script:
```python
import teslapy
tesla = teslapy.Tesla('XX@YY.ZZ')
if not tesla.authorized:
    print('Use browser to login. Page Not Found will be shown at success.')
    print('Open this URL: ' + tesla.authorization_url())
    tesla.fetch_token(authorization_response=input('Enter URL after authentication: '))
vehicles = tesla.vehicle_list()
print(vehicles[0])
tesla.close()
```

Python version is 3.9.2, running on a Debian Bullseye system.

Let me know if you need any more information or if I can check things for you or send you some debug logs etc.

**Reply by tdorssers** [reply]
I ran the above script on Ubuntu 22.04 and could not trigger the error. Packages in my virtual env: certifi==2023.5.7 charset-normalizer==3.1.0 idna==3.4 oauthlib==3.2.2 requests==2.31.0 requests-oauthlib==1.3.1 TeslaPy==2.8.0 urllib3==2.0.3 websocket-client==1.6.1
Did the same on Debian 11.7 and that triggered the error.

**Reply by tdorssers** [reply]
I have tested with Python 3.8.10 on Windows and 3.10.6 on Ubuntu with the new urllib3, both work fine. Ubuntu 21.04 comes with Python 3.9.5 which also runs fine. I believe its the combination urllib3 2.0.x and Python 3.9.2 (and some older versions as well), but I can't find out what's going wrong here. The HTTP request is identical compared to a working situation, but still the servers respond with 403. I think Debian 11 users should use `urllib3<2` or upgrade to Debian 12.
